src/ReplaySession.cc

```diff
@@ -497,7 +497,7 @@ bool ReplaySession::handle_unrecorded_cpuid_fault(
     ReplayTask* t, const StepConstraints& constraints) {
   if (t->stop_sig() != SIGSEGV || !has_cpuid_faulting() ||
       trace_in.uses_cpuid_faulting() ||
-      trapped_instruction_at(t, t->ip()) != TrappedInstruction::CPUID) {
+      x86_trapped_instruction_at(t, t->ip()) != X86TrappedInstruction::CPUID) {
```

Comment:
It's not clear what ensures we're x86-only at this point.

I think a slightly better approach would be to keep `trapped_instruction_at` and the `TrappedInstruction` enum named as is, but rename the `TrappedInstruction` enum values with `X86_` prefixes, and have `trapped_instruction_at` check `is_x86ish(t->arch())` before returning any of those values. How does that sound?

Reply:
`!has_cpuid_faulting()` will be true on non-x86 because CPUID is x86-specific, so we will only reach here on x86.

I considered something like your suggested approach, but the arm64-specific trapped instructions that I plan to introduce for timer trapping (specifically `mrs xN, cntvct_el0` etc.) take a register operand unlike the x86 instructions, so we'll need some way to return that back, and I thought that it would be best to do that via a separate function.

Reply:
FWIW `has_cpuid_faulting()` will actually die if `ARCH_SET_CPUID` ever returns success on a non-x86 system... ideally we wouldn't depend on that. So let's not: 8deb76b3042be84d80616e0643f0c83e9200c638

It's still a little obscure at this call site to require that readers know `has_cpuid_faulting()` implies x86. It's easy to imagine that might not be the case if we sooner or later support similar functionality on other architectures.

> the arm64-specific trapped instructions that I plan to introduce for timer trapping (specifically mrs xN, cntvct_el0 etc.) take a register operand unlike the x86 instructions, so we'll need some way to return that back, and I thought that it would be best to do that via a separate function.

Why not use this function? We can change it to return a struct containing the instruction type and an output register field.

Reply:
> It's still a little obscure at this call site to require that readers know has_cpuid_faulting() implies x86. It's easy to imagine that might not be the case if we sooner or later support similar functionality on other architectures.

Yeah, the current state is not ideal but if we add CPUID trapping support for other architectures this part of the code would need to be changed anyway, so I didn't think that would be a significant problem.

> Why not use this function? We can change it to return a struct containing the instruction type and an output register field.

That would be fine with me but I thought it would be slightly clearer to use the type system to express which information is available on each architecture (e.g. X86TrappedInstruction is an enum, ARMTrappedInstruction is an (enum, regno) pair) rather than having a product type for everything on every architecture. That being said it's probably not a big deal as long as we only support two architectures.

Reply:
> That would be fine with me

Let's do that then.

> but I thought it would be slightly clearer to use the type system to express which information is available on each architecture (e.g. X86TrappedInstruction is an enum, ARMTrappedInstruction is an (enum, regno) pair) rather than having a product type for everything on every architecture.

There isn't really any fundamental reason why x86 trappable instructions (which is really a misnomer; e.g. PUSHF isn't trappable, these are just instructions that are important to us) don't have an explicit output register operand. That's just an accident, more or less. Intel likes to use DX:AX for special instructions but they don't always. I think it makes total sense to have a return value struct that represents the opcode and any decoded output operand. And we should probably call it `DisassembledSpecialInstruction` or something like that.

Reply:
Done